### PR TITLE
http2: improve header single value checking

### PR DESCRIPTION
**Pull request by jasnell**
this could be a lot better but it's an improvement over what it was. Will need to play around with the implementation to come up with something that performs better.

##### Checklist

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#commit-guidelines)

##### Affected core subsystem(s)
http2

### CODE REVIEW

File: lib/internal/http2/util.js [hunk 1]
```diff
@@ -156,24 +246,25 @@ function mapToHeaders(map) {
...
-          throw new errors.Error('ERR_HTTP2_PSEUDO_HEADERS_SINGLE_VALUE');
+          throw new errors.Error('ERR_HTTP2_HEADER_SINGLE_VALUE', key);
         value = value[0];
       }
-      val = String(value);
-      assertIllegalConnectionSpecificHeader(key, val, mapToHeaders);
-      ret.unshift([key, val]);
+      ret.unshift([key, String(value)]);
     } else {
+      key = String(key).toLowerCase();
```
Comment (sebdeckers) [inline lib/internal/http2/util.js]: Headers must be lower case. Anything upper case is a protocol error. Not sure we should allow it.

File: lib/internal/http2/util.js [hunk 2]
```diff
@@ -145,6 +188,53 @@ function assertIllegalConnectionSpecificHeader(name, value, ctor) {
...
+    case HTTP2_HEADER_AUTHORITY:
+    case HTTP2_HEADER_SCHEME:
+    case HTTP2_HEADER_PATH:
+      return;
+    default:
+      throw new errors.Error('ERR_HTTP2_INVALID_PSEUDOHEADER', name);
+  }
+}
+
+// TODO(jasnell): Find a faster way to do this
```
Comment (cjihrig) [inline lib/internal/http2/util.js]: Would it be faster to load these into a Map and do a lookup here?

**Review by sebdeckers — CHANGES_REQUESTED** [review]

File: lib/internal/http2/util.js (hunk 1 above)
Comment (jasnell) [reply lib/internal/http2/util.js]: Not quite. Header field names are case insensitive and must only be transmitted in hpack as lower case. Converting to lowercase would be perfectly acceptable.

File: lib/internal/http2/util.js (hunk 2 above)
Comment (jasnell) [reply lib/internal/http2/util.js]: Possibly. I'm going to be benchmarking it tomorrow

File: lib/internal/http2/util.js (hunk 1 above)
Comment (sebdeckers) [reply lib/internal/http2/util.js]: Ah, my bad. 😌

**Review by mcollina — APPROVED** [review]
LGTM

**Comment by jasnell** [comment]
Landed

**Pull request closed without merging**